libretro-common/queues/task_queue.c

```diff
@@ -420,10 +420,9 @@ static void retro_task_threaded_wait(retro_task_condition_fn_t cond, void* data)
       retro_task_threaded_gather();
 
       slock_lock(running_lock);
-      wait = (tasks_running.front) &&
-             (!cond || cond(data));
+      wait = (tasks_running.front && !tasks_running.front->when);
       slock_unlock(running_lock);
-   } while (wait);
+   } while (wait && (!cond || cond(data)));
```

Comment:
I had to move the callback out of the lock as the callback also needs to acquire the lock to validate whether the task is still in the queue. (pthread_mutex is not re-entrant without some additional setup). Besides, it's generally not wise to have callbacks inside a lock as you don't know how long the callback will take and want to release the lock as quickly as possible.